source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+    nw_socket->event_loop = NULL;
+}
+
+static inline void s_set_socket_state(struct nw_socket *nw_socket, struct aws_socket *socket, int state) {
+    s_lock_socket_state(nw_socket);
+    // The state can only go increasing, except for LISTENING and STOPPED. They can switch between each other.
+    if (nw_socket->synced_state.state < state || (state == LISTENING && nw_socket->synced_state.state == STOPPED)) {
+        nw_socket->synced_state.state = state;
+        if (socket) {
+            socket->state = state;
+        }
+    }
+    s_unlock_socket_state(nw_socket);
+}
+
+static int s_setup_socket_params(struct nw_socket *nw_socket, const struct aws_socket_options *options) {
+    if (options->type == AWS_SOCKET_STREAM) {
+        /* if TCP, setup all the tcp options */
+        if (options->domain == AWS_SOCKET_IPV4 || options->domain == AWS_SOCKET_IPV6) {
+            // DEBUG WIP NW_PARAMETERS_DISABLE_PROTOCOL will need to be changed to use MTLS With SecItem
+            nw_socket->socket_options_to_params =
+                nw_parameters_create_secure_tcp(NW_PARAMETERS_DISABLE_PROTOCOL, ^(nw_protocol_options_t nw_options) {
+                  if (options->connect_timeout_ms) {
+                      /* this value gets set in seconds. */
+                      nw_tcp_options_set_connection_timeout(
+                          nw_options, options->connect_timeout_ms / AWS_TIMESTAMP_MILLIS);
+                  }
+
+                  // Only change default keepalive values if keepalive is true and both interval and timeout
+                  // are not zero.
+                  if (options->keepalive && options->keep_alive_interval_sec != 0 &&
+                      options->keep_alive_timeout_sec != 0) {
+                      nw_tcp_options_set_enable_keepalive(nw_options, options->keepalive);
+                      nw_tcp_options_set_keepalive_idle_time(nw_options, options->keep_alive_timeout_sec);
+                      nw_tcp_options_set_keepalive_interval(nw_options, options->keep_alive_interval_sec);
+                  }
+
+                  if (options->keep_alive_max_failed_probes) {
+                      nw_tcp_options_set_keepalive_count(nw_options, options->keep_alive_max_failed_probes);
+                  }
+
+                  if (g_aws_channel_max_fragment_size < KB_16) {
+                      nw_tcp_options_set_maximum_segment_size(nw_options, g_aws_channel_max_fragment_size);
+                  }
+                });
+        } else if (options->domain == AWS_SOCKET_LOCAL) {
+            nw_socket->socket_options_to_params =
+                nw_parameters_create_secure_tcp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
+        }
+    } else if (options->type == AWS_SOCKET_DGRAM) {
+        nw_socket->socket_options_to_params =
+            nw_parameters_create_secure_udp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
+    }
+
+    if (!nw_socket->socket_options_to_params) {
```

Comment:
I feel like we should have different error output whether or not
1. the call to `nw_parameters_*` failed
2. we fell through all the if statements and didn't even try to set up the parameters (for example, if VSOCK was the domain)